codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -160,6 +160,23 @@ private static boolean validateHeadersSentState(Http2Stream stream, Http2Headers
         return isInformational;
     }
 
+    /**
+     * According to RFC 9113, "Pseudo-header fields defined for requests MUST NOT appear in responses;
+     * pseudo-header fields defined for responses MUST NOT appear in requests."
+     * @return {@code true} if validation completed successfully, {@code false} if not
+     */
+    private boolean validatePseudoHeaders(Http2Headers headers) {
+        boolean isResponseHeaders = connection.isServer();
+        for (CharSequence name : headers.names()) {
+            if (Http2Headers.PseudoHeaderName.isPseudoHeader(name) &&
+                (isResponseHeaders && Http2Headers.PseudoHeaderName.getPseudoHeader(name).isRequestOnly()) ||
```

Comment:
Maybe make sense to modify the pseudo header names to have an array/list of all the ones request only/not request only. in order to perform the match upfront and eventually not perform any iteration/match, if not necessary.

Reply:
The iteration could be super costly especially with a large number of headers so I agree if we could skip the iteration I would. I'm thinking of an alternative to this by adding flags to the Http2 Header class itself and checking them when passing through the encoder which in my mind should be better.

Reply:
Actually on second thought, the request pseudoHeaders are actually always on the [front to be iterated](https://github.com/netty/netty/blob/b8cddab13de5a83a1be48a94e5095388aa424007/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java#L283) so it makes it more efficient to just stop going through the headers once we stop reading pseudoHeaders. I'm making changes to follow this approach.